File: problem/Add_Binary/addBinary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
const addBinary = (a, b) => {
    const max = Math.max(a.length, b.length);
    let carry = 0;
    const res = [];

    const sum = (A, B, C) => (A ^ B) ^ C;
    const carr = (A, B, C) => (A & B) | ((A ^ B) & C);

    for (let i = 0; i < max; i++) {
        const A = +(a.at(-1 - i) ?? 0);
        const B = +(b.at(-1 - i) ?? 0);

        res.push(sum(A, B, carry));
        carry = carr(A, B, carry);
    }

    if (carry) res.push(carry);

    return res.reverse().join("");
};

File: problem/Add_Binary/addBinary.cpp
class Solution
{
public:
    string addBinary(string a, string b)
    {
        string &res = (a.size() >= b.size()) ? a : b;
        int i = a.size() - 1, j = b.size() - 1;
        int k = res.size() - 1;
        int C = 0;

        while (k >= 0)
        {
            int A = i < 0 ? 0 : a[i] & 1;
            int B = j < 0 ? 0 : b[j] & 1;

            int S = (A ^ B) ^ C;
            int C_out = ((A ^ B) & C) | (A & B);

            res[k] = S + '0';

            C = C_out;
            i--;
            j--;
            k--;
        }

        return C ? (res.insert(res.begin(), '1'), res) : res;
    }
};